Lib/glyphsLib/builder/axes.py

```diff
@@ -157,8 +159,16 @@ def to_designspace_axes(self):
         ):
             axis_wanted = True
 
+        # See https://github.com/googlefonts/glyphsLib/issues/568
+        if custom_mapping and axis.tag in custom_mapping:
+            mapping = {int(k): v for k, v in custom_mapping[axis.tag].items()}
+            regularDesignLoc = axis_def.get_design_loc(regular_master)
+            # Glyphs masters don't have a user location, so we compute it by
+            # looking at the axis mapping in reverse.
```

Comment:
yes, this comment is definitely copy-pasta. What you're trying to do here is computing the user-space location of the base master, in case the Axis Mappings does not contain an explicit mapping for it.
I am wondering, should we not instead require that all minimum, default and maximum coordinates along each axes that are mapped in the new Axis Mappings be explicitly specified by the user instead of doing this linear interpolation like we do here for the default? The latter interpolation only works if the the mapping contains a min and max, and min != max. Requiring all three to be set (min, default and max, with default possibly == to either min or max) would make it easier to think about it.

Reply:
> I am wondering, should we not instead require that all minimum, default and maximum coordinates along each axes that are mapped in the new Axis Mappings be explicitly specified by the user instead of doing this linear interpolation like we do here for the default? 

I guess we could do the following:

For the min and max, we can simply use `min(mapping)`, `max(mapping)`.

For the default, we can use the "Variable Font Origin" customParameter. If it isn't present, we could use the min value for the default.

Alternatively, if there is no "Variable Font Origin" customParameter, we skip this new Axis Mappings method completely because we don't have enough information. We should then log a warning informing the user that they need to specify a "Variable font Origin" in order for it to work.

Reply:
Using the Variable Font Origin still means we'll need to convert from internal coords to user coords which may be messy.

@anthrotype I agree that it may be best to get the user to define this info themselves. We could change the Axis Mappings data structure to:
```
{
    "wght": {
        "minimum": 100, # this and maximum can be computed if needed
        "maximum": 900, 
        "default": 100,
        "mapping": {100: 100, 400: 300, 900: 900},
    "wdth": {...}
}
```

Obviously we can only do this if Georg agrees.

Reply:
> For the default, we can use the "Variable Font Origin" 

that logic is already implemented in `get_regular_master` function in this module, which returns the "regular" master (i.e. the base or default or neutral, not "regular" as the Regular style). You search that master's design location in the set of values of the axis mappings dict; if it's present, use the key (user-space location) as is; if missing, raise an error and require the user to add an explicit mapping for the default master.

Reply:
no, I don't think we need to change the data structure of Axis Mappings custom parameter. We only need to check that there are explicit mappings for the minimum, default and maximum values along each axis. Note that if we don't check and enforce that here, it will be varLib later on that will complain as it attempts to build the avar table.

The requirement that min, default and max be explicitly mapped does not mean there has to be necessarily three key:value pairs; because one of them (the default) may well be == to min or max. So the minimum number for a meaningful axis that does vary is two, with default == min or max. If there's only one mapping then the min == max == default and the variation axis is really just a point and there ought to be only one position for all the involved masters and instances for that axis.

Reply:
Thanks for the tips, I now have the following:

```
mapping = {float(k): v for k, v in custom_mapping[axis.tag].items()}
regularDesignLoc = axis_def.get_design_loc(regular_master)

reverse_mapping = {v: k for k,v in mapping.items()}
if regularDesignLoc not in reverse_mapping:
    raise KeyError("Axis Mapping must include a mapping for the default master")
regularUserLoc = reverse_mapping[regularDesignLoc]
```